websocket/src/websocket.cpp

```diff
@@ -0,0 +1,536 @@
+{
+    {"connect", LuaConnect},
+    {"disconnect", LuaDisconnect},
+    {"send", LuaSend},
+    {0, 0}
+};
+
+static void LuaInit(lua_State* L)
+{
+    int top = lua_gettop(L);
+
+    // Register lua names
+    luaL_register(L, MODULE_NAME, Websocket_module_methods);
+
+#define SETCONSTANT(_X) \
+            lua_pushnumber(L, (lua_Number) _X); \
+            lua_setfield(L, -2, #_X);
+
+        SETCONSTANT(EVENT_CONNECTED);
+        SETCONSTANT(EVENT_DISCONNECTED);
+        SETCONSTANT(EVENT_MESSAGE);
+        SETCONSTANT(EVENT_ERROR);
+
+#undef SETCONSTANT
+
+    lua_pop(L, 1);
+    assert(top == lua_gettop(L));
+}
+
+static dmExtension::Result WebsocketAppInitialize(dmExtension::AppParams* params)
+{
+    g_Websocket.m_BufferSize = dmConfigFile::GetInt(params->m_ConfigFile, "websocket.buffer_size", 64 * 1024);
+    g_Websocket.m_Timeout = dmConfigFile::GetInt(params->m_ConfigFile, "websocket.socket_timeout", 500 * 1000);
+    g_Websocket.m_Connections.SetCapacity(4);
+    g_Websocket.m_Channel = 0;
+    g_Websocket.m_Pool = 0;
+
+    dmConnectionPool::Params pool_params;
+    pool_params.m_MaxConnections = dmConfigFile::GetInt(params->m_ConfigFile, "websocket.max_connections", 2);
+    dmConnectionPool::Result result = dmConnectionPool::New(&pool_params, &g_Websocket.m_Pool);
+
+    if (dmConnectionPool::RESULT_OK != result)
+    {
+        dmLogError("Failed to create connection pool: %d", result);
+    }
+
+// We can do without the channel, it will then fallback to the dmSocket::GetHostname (as opposed to dmDNS::GetHostname)
+#if defined(HAVE_WSLAY)
+    dmDNS::Result dns_result = dmDNS::NewChannel(&g_Websocket.m_Channel);
+
+    if (dmDNS::RESULT_OK != dns_result)
+    {
+        dmLogError("Failed to create connection pool: %d", dns_result);
+    }
+#endif
+
+    g_Websocket.m_Initialized = 1;
+    if (!g_Websocket.m_Pool)
+    {
+        if (!g_Websocket.m_Pool)
+        {
+            dmLogInfo("pool is null!");
+            dmConnectionPool::Delete(g_Websocket.m_Pool);
+        }
+
+        dmLogInfo("%s extension not initialized", MODULE_NAME);
+        g_Websocket.m_Initialized = 0;
+    }
+
+    return dmExtension::RESULT_OK;
+}
+
+static dmExtension::Result WebsocketInitialize(dmExtension::Params* params)
+{
+    if (!g_Websocket.m_Initialized)
+        return dmExtension::RESULT_OK;
+
+    LuaInit(params->m_L);
+    dmLogInfo("Registered %s extension", MODULE_NAME);
+
+    return dmExtension::RESULT_OK;
+}
+
+static dmExtension::Result WebsocketAppFinalize(dmExtension::AppParams* params)
+{
+
+    dmConnectionPool::Shutdown(g_Websocket.m_Pool, dmSocket::SHUTDOWNTYPE_READWRITE);
+    return dmExtension::RESULT_OK;
+}
+
+static dmExtension::Result WebsocketFinalize(dmExtension::Params* params)
+{
+    return dmExtension::RESULT_OK;
+}
+
+static dmExtension::Result WebsocketOnUpdate(dmExtension::Params* params)
+{
+    uint32_t size = g_Websocket.m_Connections.Size();
+
+    for (uint32_t i = 0; i < size; ++i)
```

Comment:
The state handling.
Upon error, the status flag is set, and the connection is closed.
The cleanup happens next frame, for the `STATE_DISCONNECTED` connections.